### PR TITLE
tamarin-prover: build SAPIC as default

### PR DESCRIPTION
**Pull request by lordqwerty**
Added "with-sapic" option to now allow the Stateful Applied Pi Calculus
plugin to be built.

### Have you:

- [X] Followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-science/blob/master/.github/CONTRIBUTING.md) document?
- [X] Checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-science/pulls) for the same update/change?
- [X] Run `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting) and corrected all errors?
- [X] Built your formula locally prior to submission with `brew install <formula>`?

### New formula: have you

- [ ] Written a sensible test? (The best test is to compile and run a sample program.)

### Updates to existing formula: have you

- [ ] Removed the `revision` line, if any, when bumping the version number?
- [ ] Added/bumped the `revision` number if the changes affect the precompiled bottles?
- [X] Not altered the `bottle` section?
- [X] Checked that the tests still pass?

### CODE REVIEW

File: tamarin-prover.rb [hunk 1]
```diff
@@ -12,15 +12,26 @@ class TamarinProver < Formula
     sha256 "2b2cee8f7439698171ecc1fc31e171439a711e7831aad035be64e34adb478e91" => :yosemite
   end
 
+  option "with-ocaml", "Enable Stateful Applied Pi Calculus plugin (SAPIC)"
```
Comment (katrielalex) [inline tamarin-prover.rb]: Should this be "with-sapic"?

File: tamarin-prover.rb [hunk 2]
```diff
@@ -12,15 +12,26 @@ class TamarinProver < Formula
...
+
   depends_on "haskell-stack" => :build
   depends_on "maude"
   depends_on "graphviz"
+  depends_on "ocaml" => :optional
```
Comment (katrielalex) [inline tamarin-prover.rb]: My intuition says this should be 
```
depends_on "ocaml" => : build if build.with?("sapic")
```
or whatever the syntax is.

Comment (lordqwerty) [inline tamarin-prover.rb]: Changed. Many thanks.

**Comment by lordqwerty** [comment]
@jonchang can you confirm this is a time out issue like with PR https://github.com/Homebrew/homebrew-science/pull/5176 ? Sorry to ask you directly but since you looked at formula originally I thought you best to ask.

**Comment by mpdehnel** [comment]
Yup, looks identical to the Linuxbrew issues with Tamarin-Prover (unsurprisingly). Think this is because Homebrew doesn't have Linux bottles for Haskell, iirc.

**Comment by iMichka** [comment]
The Linux error is due to `libtecla` (one of the dependencies of `maude`) failing to build. We can ignore the Linux build for the moment. Though if somebody wants to work in this, it would be great.

File: tamarin-prover.rb [hunk 3]
```diff
@@ -12,15 +12,26 @@ class TamarinProver < Formula
...
+  depends_on "ocaml" => :build if build.with?("sapic")
   depends_on :macos => :mountain_lion
 
   # doi "10.1109/CSF.2012.25"
   # tag "security"
 
   def install
+    if build.with?("sapic")
+      ENV.deparallelize do
+        system "make", "sapic"
```
Comment (jonchang) [inline tamarin-prover.rb]: This should be moved into the other block. 
`system "make", "sapic" if build.with("sapic")`

File: tamarin-prover.rb [hunk 4]
```diff
@@ -12,15 +12,26 @@ class TamarinProver < Formula
...
+      end
+
+      bin.install ".brew_home/.local/bin/sapic"
```
Comment (jonchang) [inline tamarin-prover.rb]: This should be merged with the other `bin.install` command below. Perhaps something like `bin.install Dir[".brew_home/.local/bin/*"]`?

**Comment by jonchang** [comment]
If anyone wants to work on the `libtecla` Linux issue feel free to ping me or other Linuxbrew maintainers for help.

**Comment by lordqwerty** [comment]
Changed as request and now builds SAPIC plug-in by default which makes more sense. Revision added so new bottles are built to contain both binaries. 

Many thanks for the feedback and support. 👍

**Comment by mpdehnel** [comment]
Would it be sensible to add an option such as `--without-sapic` in case there's an issue / someone doesn't want to install OCaml?

**Comment by lordqwerty** [comment]
@mpdehnel as requested a build without option for sapic. This should build sapic binary as default and therefore reduces the options still I believe.

**Comment by lordqwerty** [comment]
I don't understand whats is causing the error:
```
==> Cleaning
Error: File exists - /usr/local/opt/tamarin-prover/.brew
```

Any ideas?

**Comment by mpdehnel** [comment]
@lordqwerty: Kevin says: 
>The build error was a stack one where it tries to expand ghc multiple times; expand the download ghc archive

Don't know if this helps :-)

File: tamarin-prover.rb [hunk 5]
```diff
@@ -12,7 +13,10 @@ class TamarinProver < Formula
     sha256 "2b2cee8f7439698171ecc1fc31e171439a711e7831aad035be64e34adb478e91" => :yosemite
   end
 
-  depends_on "haskell-stack" => :build
+  option "without-sapic", "Compile without SAPIC plug-in"
+
+  depends_on "haskell-stack"  =>   :build
```
Comment (ilovezfs) [inline tamarin-prover.rb]: Please keep the white space as just one space on each side of the hash rockets.

Comment (lordqwerty) [inline tamarin-prover.rb]: Done. 👍

**Comment by lordqwerty** [comment]
Just to let you all know I have worked with the Linuxbrew team and now have `libtecla` building on Linux builds. Relevant PRs:

- https://github.com/Linuxbrew/brew/pull/320
- https://github.com/Linuxbrew/homebrew-core/pull/2238

Now the builds just timeout: `The job exceeded the maximum time limit for jobs, and has been terminated.` Not much we can do about this I guess.

**Comment by iMichka** [comment]
Maude pull request has been merged. You can now rebase this.

**Comment by lordqwerty** [comment]
`command umask 022; PATH="/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:$PATH"; eval "$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib)"; brew install patchelf && brew tap homebrew/dupes && brew tap linuxbrew/xorg && brew test-bot took more than 120 minutes to run` 

Takes too long to build on CircleCI :(

**Comment by lordqwerty** [comment]
Linux builds now! Sorry for the delay in fixes.

**Comment by lordqwerty** [comment]
OK. I'm not touching this any further. Please merge. Sorry for the long PR.

**Comment by jonchang** [comment]
Thanks for all your work on this @lordqwerty and fixing formulae up for Linuxbrew! This has been merged.

**Pull request closed without merging**